File: codes/ruby/chapter06/answer_A38.rb
# 入力
d, n = gets.split.map(&:to_i)
l, r, h = [], [], []
n.times do
  li, ri, hi = gets.split.map(&:to_i)
  l << li
  r << ri
  h << hi
end

# 配列の初期化（1 日は 24 時間）
lim = Array.new(d + 1, 24)

# 上限値を求める
n.times do |i|
  (l[i]..r[i]).each do |j|
    lim[j] = [lim[j], h[i]].min
  end
end

# 答えを出力
answer = lim[1..d].sum
puts answer

File: codes/go/chapter06/answer_A38.go
package main

import (
	"fmt"
)

func main() {
	var D, N int
	fmt.Scan(&D, &N)

	L := make([]int, N+1)
	R := make([]int, N+1)
	H := make([]int, N+1)

	for i := 1; i <= N; i++ {
		fmt.Scan(&L[i], &R[i], &H[i])
	}

	LIM := make([]int, D+1)
	for i := 1; i <= D; i++ {
		LIM[i] = 24
	}

	for i := 1; i <= N; i++ {
		for j := L[i]; j <= R[i]; j++ {
			if LIM[j] > H[i] {
				LIM[j] = H[i]
			}
		}
	}

	var Answer int
	for i := 1; i <= D; i++ {
		Answer += LIM[i]
	}
	fmt.Println(Answer)
}
